### PR TITLE
Use the same session creation logic in both Panache Session Operation and @Transactional

### LINKED ISSUE #54537: `@Transactional` does not work in Hibernate Reactive with Panache
Apparently, `HibernateReactiveRecorder.getSession` and `SessionOperations.getSession` don't have the same set of opened sessions, which means Hibernate Reactive with Panache can't use `@Transactional` ATM, because it doesn't find any opened session.

We need to make sure this works.

### PR DESCRIPTION
**Pull request by lucamolteni**
Collapse duplicated reuse-or-create logic between SessionOperations and HibernateReactiveRecorder by parameterizing the context key.

Expose Vert.x transaction through Panache.currentTransaction() when running under @Transactional via VertxTransactionWrapper.

Split test endpoints and test classes into abstract base + two subclasses (WithTransaction and Transactional) to test both annotations without shared DB state. Same for Kotlin.

Alternative implementation of https://github.com/quarkusio/quarkus/pull/54566

* Fixes https://github.com/quarkusio/quarkus/issues/54537

### CODE REVIEW

File: extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/VertxTransactionWrapper.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    @Override
+    public void markForRollback() {
+        vertxTransaction.rollback();
+    }
+
+    @Override
+    public boolean isMarkedForRollback() {
+        return false;
```
Comment (lucamolteni) [inline extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/VertxTransactionWrapper.java]: I'm not sure about this

**Comment by lucamolteni** [comment]
@FroMage  I've used the same state for @ Transactional and Session Operation, removing a lot of duplicated code between the two. I've also exposed the current transaction in both WithTransaction and @ Transactional by adding a current transaction method to Session Operations. I've also created a super class in the test so that we can test both implementation, when we decide to remove the other annotations we can remove it. Since transaction is created lazy though I had to change the assertion order because we couldn't assert transaction != null as first element of the chain

**Comment by FroMage** [comment]
Sounds great. But due to past-me (that horrible horrible person) I'm very late working on Monday's demo so I won't have time to review until Tuesday, sorry 😬

**Comment by lucamolteni** [comment]
> Sounds great. But due to past-me (that horrible horrible person) I'm very late working on Monday's demo so I won't have time to review until Tuesday, sorry 😬

no problem take your time!

**Comment by FroMage** [comment]
The test failures do seem related.

**Comment by cescoffier** [comment]
🤖 Automatically associated with the **WG - Quarkus Data** Working Group based on AI classification.

**Comment by lucamolteni** [comment]
@FroMage note that out of +2,760 -2,188 the fix is much simpler and it's mostly about SessionOperations. 

The rest of the PR is to parametrize Panache tests to run on @ Transactional  and @ WithTransaction

I'm not convinced yet about this. I also used Claude for the refactor and I can see manual reviewing is complicated, so sorry in advance for that. We might decide to go for the fix and migrate all Panache's test to @ Transactional and just keep a simple SmokeTest for @ WithTransaction in order to see everything still works.

**Comment by FroMage** [comment]
Let's migrate all tests to `@Transactional` and keep a single test with `@WithTransaction` just to make sure it works.

File: extensions/panache/hibernate-reactive-panache-common/deployment/src/main/java/io/quarkus/hibernate/reactive/panache/common/deployment/PanacheJpaCommonResourceProcessor.java
Comment (FroMage) [inline extensions/panache/hibernate-reactive-panache-common/deployment/src/main/java/io/quarkus/hibernate/reactive/panache/common/deployment/PanacheJpaCommonResourceProcessor.java]: That comment also needs to be updated.

File: extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java [hunk 2]
```diff
@@ -37,41 +35,8 @@ public final class SessionOperations {
...
-                .unwrap(SESSION_FACTORY_MAP.getValue(persistenceUnitName));
-        return new BaseKey<>(Mutiny.Session.class, implementor.getUuid());
-    }
-
-    private static Key<Mutiny.StatelessSession> createStatelessSessionKey(String persistenceUnitName) {
-        Implementor implementor = (Implementor) ClientProxy
-                .unwrap(SESSION_FACTORY_MAP.getValue(persistenceUnitName));
-        return new BaseKey<>(Mutiny.StatelessSession.class, implementor.getUuid());
-    }
+    private static final OpenedSessionsState<Mutiny.Session> SESSIONS = HibernateReactiveRecorder.OPENED_SESSIONS_STATE;
```
Comment (FroMage) [inline extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java]: I'd prefer to directly refer to `HibernateReactiveRecorder.OPENED_SESSIONS_STATE` rather than keep a copy in a separate field. This could confuse people into thinking they're different.

File: extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java [hunk 3]
```diff
@@ -322,38 +281,22 @@ public static Uni<Mutiny.Session> getSession(String persistenceUnitName) {
...
-            } else {
-                throw new IllegalStateException("No current Mutiny.Session found"
-                        + "\n\t- no reactive session was found in the Vert.x context and the context was not marked to open a new session lazily"
-                        + "\n\t- a session is opened automatically for JAX-RS resource methods annotated with an HTTP method (@GET, @POST, etc.); inherited annotations are not taken into account"
-                        + "\n\t- you may need to annotate the business method with @WithSession or @WithTransaction");
-            }
+            throw new IllegalStateException("No current Mutiny.Session found"
+                    + "\n\t- no reactive session was found in the Vert.x context and the context was not marked to open a new session lazily"
+                    + "\n\t- a session is opened automatically for JAX-RS resource methods annotated with an HTTP method (@GET, @POST, etc.); inherited annotations are not taken into account"
+                    + "\n\t- you may need to annotate the business method with @WithSession or @WithTransaction");
```
Comment (FroMage) [inline extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java]:
```suggestion
                    + "\n\t- you may need to annotate the business method with @Transactional, @WithSession or @WithTransaction");
```

File: extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java [hunk 4]
```diff
@@ -468,24 +417,10 @@ public static Context vertxContext() {
...
-        if (currentStateless != null && currentStateless.isOpen()) {
-            LOG.debugf("Closing opened stateless session for Persistence Unit '%s'", persistenceUnitName);
-            return currentStateless.close().eventually(() -> context.removeLocal(statelessKey));
-        }
-        return Uni.createFrom().voidItem();
+        return SESSIONS.closeSession(context, persistenceUnitName)
+                .chain(() -> STATELESS_SESSIONS.closeSession(context, persistenceUnitName));
     }
 
     static void clear() {
```
Comment (FroMage) [inline extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java]: If this doesn't do anything anymore, let's remove it.

File: extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java [hunk 5]
```diff
@@ -452,6 +384,23 @@ public static Mutiny.StatelessSession getCurrentStatelessSession(String persiste
      * @throws IllegalStateException If no vertx context is found or is not a safe context as mandated by the
      *         {@link VertxContextSafetyToggle}
      */
+    public static Uni<Mutiny.Transaction> currentTransaction() {
+        return getSession().map(session -> {
+            Mutiny.Transaction tx = session.currentTransaction();
+            if (tx != null) {
+                return tx;
+            }
+            Future<? extends SqlConnection> connectionFuture = TransactionalContextPool.getCurrentConnectionFromVertxContext();
```
Comment (FroMage) [inline extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java]: I would much rather `session.currentTransaction()` also "noticed" that we do have a TX open, because we're moving the issue here. Anybody using that API will not notice that we have a pending TX open.

/cc @DavideD

File: extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java [hunk 6]
```diff
@@ -452,6 +384,23 @@ public static Mutiny.StatelessSession getCurrentStatelessSession(String persiste
...
+            if (connectionFuture != null && connectionFuture.succeeded()) {
```
Comment (FroMage) [inline extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java]: This is confusing: we're in a reactive method returning a `Uni` and yet we're expecting that a `Future` has already resolved, rather than wait on it as part of the reactive chain.
Is there an underlying assumption that it will always have already resolved?

Or is it perhaps that if it hasn't resolved then we possibly can't have a current TX?

File: extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/VertxTransactionWrapper.java (hunk 1 above)
Comment (FroMage) [reply extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/VertxTransactionWrapper.java]: That's probably because it's wrong. We should really make sure that HR's transactions and ours are one and the same, and agree on the current state of the TX.

File: extensions/panache/hibernate-reactive-panache-kotlin/runtime/src/main/kotlin/io/quarkus/hibernate/reactive/panache/kotlin/Panache.java [hunk 7]
```diff
@@ -135,6 +135,6 @@ public static Uni<Void> flush() {
      * @see Panache#withTransaction(Supplier)
      */
     public static Uni<Mutiny.Transaction> currentTransaction() {
-        return getSession().map(session -> session.currentTransaction());
```
Comment (FroMage) [inline extensions/panache/hibernate-reactive-panache-kotlin/runtime/src/main/kotlin/io/quarkus/hibernate/reactive/panache/kotlin/Panache.java:138]: While I'm all for simplifications such as this, we also have to make sure that `session.currentTransaction()` returns the same TX as `SessionOperations.currentTransaction()`.

/cc @DavideD

File: integration-tests/hibernate-reactive-panache/src/test/java/io/quarkus/it/panache/reactive/PanacheFunctionalityTest.java [hunk 8]
```diff
@@ -251,12 +251,14 @@ public void testReactiveTransactional(UniAsserter asserter) {
         asserter.assertEquals(() -> reactiveTransactional(), 1l);
     }
 
-    @WithTransaction
+    // We shall assert tx is not null after the first operation
+    // As the session is created lazily
+    @Transactional
```
Comment (FroMage) [inline integration-tests/hibernate-reactive-panache/src/test/java/io/quarkus/it/panache/reactive/PanacheFunctionalityTest.java:258]: We should probably keep one `@WithTransaction` test that checks that initially the TX is not `null` as that's defined behaviour.

Looks like we also need a `@Transactional` test that verifies that the TX is `null` before any operation and not afterwards?

**Review by FroMage — CHANGES_REQUESTED** [review]

**Comment by FroMage** [comment]
BTW, I think this PR is much cleaner than what we had before.

File: extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java (hunk 5 above)
Comment (DavideD) [reply extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java]: Ok, I will start to look into it

File: extensions/panache/hibernate-reactive-panache-kotlin/runtime/src/main/kotlin/io/quarkus/hibernate/reactive/panache/kotlin/Panache.java (hunk 7 above)
Comment (DavideD) [reply extensions/panache/hibernate-reactive-panache-kotlin/runtime/src/main/kotlin/io/quarkus/hibernate/reactive/panache/kotlin/Panache.java:138]: Is there already a test for this situation, or do we need to create one?

File: integration-tests/hibernate-reactive-panache/src/test/java/io/quarkus/it/panache/reactive/PanacheFunctionalityTest.java (hunk 8 above)
Comment (lucamolteni) [reply integration-tests/hibernate-reactive-panache/src/test/java/io/quarkus/it/panache/reactive/PanacheFunctionalityTest.java:258]: Added the tx null assertion first and a new test https://github.com/quarkusio/quarkus/pull/54683/changes/7f6a064df84aaf3ffa75e0522aa2595d97671007

File: extensions/panache/hibernate-reactive-panache-kotlin/runtime/src/main/kotlin/io/quarkus/hibernate/reactive/panache/kotlin/Panache.java (hunk 7 above)
Comment (FroMage) [reply extensions/panache/hibernate-reactive-panache-kotlin/runtime/src/main/kotlin/io/quarkus/hibernate/reactive/panache/kotlin/Panache.java:138]: I don't think there's a test, no, but it should be easy:

- make a method `@Transactional`
- use an injected `Mutiny.Session`
- persist an entity
- verify that `session.currentTransaction()` is `null` even though there is a transaction in progress.

Comment (FroMage) [reply extensions/panache/hibernate-reactive-panache-kotlin/runtime/src/main/kotlin/io/quarkus/hibernate/reactive/panache/kotlin/Panache.java:138]: This particular issue is fixed by https://github.com/quarkusio/quarkus/pull/54890

File: extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java
Comment (FroMage) [inline extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java]: With https://github.com/quarkusio/quarkus/pull/54890 we can rely on `session.currentTransaction()` again.

**Review by FroMage — CHANGES_REQUESTED** [review]

File: extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java
Comment (lucamolteni) [reply extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java]: yes I'm verifying it works

Comment (FroMage) [reply extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java]: It better work because I merged it anyway 😬😱

Comment (lucamolteni) [reply extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java]: it should be ok and ready for review https://github.com/lucamolteni/quarkus/actions/runs/27694689315

**Comment by lucamolteni** [comment]
> That's probably because it's wrong. We should really make sure that HR's transactions and ours are one and the same, and agree on the current state of the TX.

Just to be clear: HR won't rollback the external transaction, it will just mark it. I'm rollbacking it in the TransactionalInterceptorBase

**Comment by FroMage** [comment]
> Just to be clear: HR won't rollback the external transaction, it will just mark it. I'm rollbacking it in the TransactionalInterceptorBase

Yes, but they have to agree on whether or not it should be rolled back.

File: extensions/hibernate-reactive/runtime/src/main/java/io/quarkus/hibernate/reactive/runtime/transaction/HibernateActionsStrategy.java [hunk 9]
```diff
@@ -51,4 +54,18 @@ public Uni<Void> afterCommit(Context context) {
...
 
     }
+
+    @Override
+    public boolean isMarkedForRollback(Context context) {
+        Optional<String> optPersistenceUnitName = getPersistenceUnitName(context);
+        if (optPersistenceUnitName.isEmpty()) {
+            return false;
+        }
+        return HibernateReactiveRecorder.OPENED_SESSIONS_STATE
```
Comment (FroMage) [inline extensions/hibernate-reactive/runtime/src/main/java/io/quarkus/hibernate/reactive/runtime/transaction/HibernateActionsStrategy.java]: Doesn't this miss stateless sessions?

**Review by FroMage — CHANGES_REQUESTED** [review]

File: extensions/hibernate-reactive/runtime/src/main/java/io/quarkus/hibernate/reactive/runtime/transaction/HibernateActionsStrategy.java (hunk 9 above)
Comment (lucamolteni) [reply extensions/hibernate-reactive/runtime/src/main/java/io/quarkus/hibernate/reactive/runtime/transaction/HibernateActionsStrategy.java]: Good catch, I missed that

File: extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java (hunk 6 above)
Comment (lucamolteni) [reply extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java]: Solved with HR support for external TX

**Review by FroMage — APPROVED** [review]

**Comment by lucamolteni** [comment]
I think we can merge this

**Review by yrodiere — APPROVED** [review]

**Comment by cescoffier** [comment]
🤖 Automatically associated with the **WG - Quarkus Data** Working Group based on AI classification.
